File: mirror/tier2-dart/showcase/meta_generation.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Metaprogramming
// Module: meta_generation.dart

Function generateAdder(int n) {
  return (int x) => x + n;
}

Function generateMultiplier(int n) {
  return (int x) => x * n;
}

Function composeGenerators(List<Function> generators) {
  return (int x) {
    int result = x;
    for (final gen in generators) {
      result = (gen as Function)(result) as int;
    }
    return result;
  };
}

Function metaFactory(String op, int value) {
  switch (op) {
    case 'add': return generateAdder(value);
    case 'mul': return generateMultiplier(value);
    default: return (int x) => x;
  }
}

void main() {
  final add10 = generateAdder(10);
  final mul5 = generateMultiplier(5);
  final composed = composeGenerators([
    generateAdder(5),
    generateMultiplier(2),
  ]);
  
  final result = <String, dynamic>{
    'add10to20': add10(20),
    'mul5to10': mul5(10),
    'composed30': composed(10),
    'factoryAdd': metaFactory('add', 100)(50),
    'factoryMul': metaFactory('mul', 3)(7),
  };
  print(result);
}

File: mirror/tier2-python/showcase/meta_generation.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Metaprogramming
Module: meta_generation.py
Purpose: Demonstrate code generation patterns
"""

from typing import Callable


def generate_adder(n: int) -> Callable:
    """Generate an adder function"""
    def adder(x: int) -> int:
        return x + n
    return adder


def generate_multiplier(n: int) -> Callable:
    """Generate a multiplier function"""
    def multiplier(x: int) -> int:
        return x * n
    return multiplier


def compose_generators(*generators) -> Callable:
    """Compose multiple generator functions"""
    def composed(x):
        result = x
        for gen in generators:
            result = gen(result)
        return result
    return composed


def meta_factory(op: str, value: int) -> Callable:
    """Factory for generating operations"""
    if op == "add":
        return generate_adder(value)
    elif op == "mul":
        return generate_multiplier(value)
    else:
        return lambda x: x


add_10 = generate_adder(10)
mul_5 = generate_multiplier(5)
composed = compose_generators(
    generate_adder(5),
    generate_multiplier(2)
)

result = {
    "add_10_to_20": add_10(20),
    "mul_5_to_10": mul_5(10),
    "composed_30": composed(10),
    "factory_add": meta_factory("add", 100)(50),
    "factory_mul": meta_factory("mul", 3)(7),
}

if __name__ == "__main__":
    print(result)
